File: internal/models/agent.go
package models

import (
	"encoding/json"
	"time"
)

type AgentDocument struct {
	Key       string          `json:"key"`
	Domain    string          `json:"domain"`
	Payload   json.RawMessage `json:"payload"`
	UpdatedAt time.Time       `json:"updated_at"`
}

type AgentSnapshot struct {
	Settings      AgentSettings          `json:"settings"`
	DirectInput   AgentDirectInputConfig `json:"direct_input"`
	WeComMenu     AgentWeComMenuSnapshot `json:"wecom_menu"`
	Push          AgentPushSnapshot      `json:"push"`
	Conversations []AgentConversation    `json:"conversations"`
	Memory        AgentMemorySnapshot    `json:"memory"`
	TopicSummary  AgentTopicSnapshot     `json:"topic_summary"`
	Writing       AgentWritingSnapshot   `json:"writing"`
	Market        AgentMarketSnapshot    `json:"market"`
	Evolution     AgentEvolutionSnapshot `json:"evolution"`
	UpdatedAt     time.Time              `json:"updated_at"`
}

type AgentSettings struct {
	RuntimeMode          string                `json:"runtime_mode"`
	DefaultLLMProviderID string                `json:"default_llm_provider_id"`
	LLMProviders         []AgentLLMProvider    `json:"llm_providers"`
	WeCom                AgentWeComConfig      `json:"wecom"`
	Terminal             AgentTerminalConfig   `json:"terminal"`
	Evolution            AgentEvolutionConfig  `json:"evolution"`
	STT                  AgentSpeechConfig     `json:"stt"`
	SearchEngines        []AgentSearchProvider `json:"search_engines"`
	Memory               AgentMemoryConfig     `json:"memory"`
	MD2Img               AgentMD2ImgConfig     `json:"md2img"`
	UpdatedAt            time.Time             `json:"updated_at"`
}

type AgentLLMProvider struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	Type      string `json:"type"`
	BaseURL   string `json:"base_url,omitempty"`
	APIKey    string `json:"api_key,omitempty"`
	Model     string `json:"model,omitempty"`
	ChatPath  string `json:"chat_path,omitempty"`
	TimeoutMS int    `json:"timeout_ms,omitempty"`
}

type AgentSearchProvider struct {
	ID      string         `json:"id"`
	Name    string         `json:"name"`
	Type    string         `json:"type"`
	Enabled bool           `json:"enabled"`
	Config  map[string]any `json:"config,omitempty"`
}

type AgentSpeechConfig struct {
	Provider  string `json:"provider,omitempty"`
	Command   string `json:"command,omitempty"`
	Enabled   bool   `json:"enabled"`
	TimeoutMS int    `json:"timeout_ms,omitempty"`
}

type AgentMemoryConfig struct {
	Enabled              bool `json:"enabled"`
	CompactEveryRounds   int  `json:"compact_every_rounds,omitempty"`
	CompactMaxBatchSize  int  `json:"compact_max_batch_size,omitempty"`
	SummaryTopK          int  `json:"summary_top_k,omitempty"`
	RawRefLimit          int  `json:"raw_ref_limit,omitempty"`
	RawRecordLimit       int  `json:"raw_record_limit,omitempty"`
	WindowTimeoutSeconds int  `json:"window_timeout_seconds,omitempty"`
	WindowMaxTurns       int  `json:"window_max_turns,omitempty"`
}

type AgentMD2ImgConfig struct {
	Enabled   bool   `json:"enabled"`
	Mode      string `json:"mode,omitempty"`
	Command   string `json:"command,omitempty"`
	OutputDir string `json:"output_dir,omitempty"`
	TimeoutMS int    `json:"timeout_ms,omitempty"`
}

type AgentWeComConfig struct {
	CorpID              string `json:"corp_id,omitempty"`
	CorpSecret          string `json:"corp_secret,omitempty"`
	AgentID             string `json:"agent_id,omitempty"`
	BaseURL             string `json:"base_url,omitempty"`
	BridgeURL           string `json:"bridge_url,omitempty"`
	BridgeToken         string `json:"bridge_token,omitempty"`
	BridgeStreamEnabled bool   `json:"bridge_stream_enabled,omitempty"`
	AudioDir            string `json:"audio_dir,omitempty"`
	TextMaxBytes        int    `json:"text_max_bytes,omitempty"`
	Enabled             bool   `json:"enabled"`
}

type AgentTerminalConfig struct {
	Enabled   bool   `json:"enabled"`
	CWD       string `json:"cwd,omitempty"`
	TimeoutMS int    `json:"timeout_ms,omitempty"`
}

type AgentEvolutionConfig struct {
	Command         string                      `json:"command,omitempty"`
	CWD             string                      `json:"cwd,omitempty"`
	TimeoutMS       int                         `json:"timeout_ms,omitempty"`
	CodexModel      string                      `json:"codex_model,omitempty"`
	CodexReasoning  string                      `json:"codex_reasoning,omitempty"`
	MaxFixAttempts  int                         `json:"max_fix_attempts,omitempty"`
	TestCommands    []AgentEvolutionTestCommand `json:"test_commands,omitempty"`
	AutoCommit      bool                        `json:"auto_commit,omitempty"`
	AutoPush        bool                        `json:"auto_push,omitempty"`
	PushRemote      string                      `json:"push_remote,omitempty"`
	PushBranch      string                      `json:"push_branch,omitempty"`
	StructureReview bool                        `json:"structure_review,omitempty"`
}

type AgentEvolutionTestCommand struct {
	Name      string `json:"name"`
	Command   string `json:"command"`
	TimeoutMS int    `json:"timeout_ms,omitempty"`
}

type AgentDirectInputConfig struct {
	Version   int                    `json:"version"`
	Rules     []AgentDirectInputRule `json:"rules"`
	UpdatedAt time.Time              `json:"updated_at"`
}

type AgentDirectInputRule struct {
	ID         string `json:"id"`
	Name       string `json:"name"`
	Pattern    string `json:"pattern"`
	TargetText string `json:"target_text"`
	MatchMode  string `json:"match_mode"`
	Enabled    bool   `json:"enabled"`
}

type AgentWeComMenuSnapshot struct {
	Config           AgentWeComMenuConfig    `json:"config"`
	RecentEvents     []AgentWeComEventRecord `json:"recent_events"`
	PublishPayload   map[string]any          `json:"publish_payload,omitempty"`
	ValidationErrors []string                `json:"validation_errors,omitempty"`
}

type AgentWeComMenuConfig struct {
	Version         int                `json:"version"`
	Buttons         []AgentWeComButton `json:"buttons"`
	UpdatedAt       time.Time          `json:"updated_at"`
	LastPublishedAt *time.Time         `json:"last_published_at,omitempty"`
}

type AgentWeComButton struct {
	ID           string             `json:"id"`
	Name         string             `json:"name"`
	Key          string             `json:"key"`
	Enabled      bool               `json:"enabled"`
	DispatchText string             `json:"dispatch_text"`
	SubButtons   []AgentWeComButton `json:"sub_buttons,omitempty"`
}

type AgentWeComEventRecord struct {
	ID                string    `json:"id"`
	EventType         string    `json:"event_type"`
	EventKey          string    `json:"event_key"`
	FromUser          string    `json:"from_user"`
	ToUser            string    `json:"to_user"`
	AgentID           string    `json:"agent_id,omitempty"`
	MatchedButtonID   string    `json:"matched_button_id,omitempty"`
	MatchedButtonName string    `json:"matched_button_name,omitempty"`
	DispatchText      string    `json:"dispatch_text,omitempty"`
	Status            string    `json:"status"`
	Error             string    `json:"error,omitempty"`
	ReceivedAt        time.Time `json:"received_at"`
}

type AgentWeComInboundResult struct {
	Record       AgentWeComEventRecord `json:"record"`
	ResponseText string                `json:"response_text,omitempty"`
	FromUser     string                `json:"from_user,omitempty"`
	ToUser       string                `json:"to_user,omitempty"`
}

type AgentPushSnapshot struct {
	Users     []AgentPushUser `json:"users"`
	Tasks     []AgentPushTask `json:"tasks"`
	UpdatedAt time.Time       `json:"updated_at"`
}

type AgentPushUser struct {
	ID        string    `json:"id"`
	Name      string    `json:"name"`
	WeComUser string    `json:"wecom_user,omitempty"`
	Enabled   bool      `json:"enabled"`
	UpdatedAt time.Time `json:"updated_at"`
}

type AgentPushTask struct {
	ID        string     `json:"id"`
	Name      string     `json:"name"`
	UserID    string     `json:"user_id"`
	Text      string     `json:"text"`
	IntervalM int        `json:"interval_minutes"`
	Enabled   bool       `json:"enabled"`
	LastRunAt *time.Time `json:"last_run_at,omitempty"`
	NextRunAt *time.Time `json:"next_run_at,omitempty"`
	UpdatedAt time.Time  `json:"updated_at"`
}

type AgentConversation struct {
	ID        string         `json:"id"`
	SessionID string         `json:"session_id"`
	Input     string         `json:"input"`
	Resolved  string         `json:"resolved,omitempty"`
	Response  string         `json:"response"`
	Status    string         `json:"status"`
	Metadata  map[string]any `json:"metadata,omitempty"`
	CreatedAt time.Time      `json:"created_at"`
}

type AgentConversationRequest struct {
	SessionID string `json:"session_id,omitempty"`
	Input     string `json:"input"`
	Actor     string `json:"actor,omitempty"`
}

type AgentTopicSnapshot struct {
	ActiveProfileID string                  `json:"active_profile_id"`
	Profiles        []AgentTopicProfile     `json:"profiles"`
	Runs            []AgentTopicRun         `json:"runs"`
	SentLog         []AgentTopicSentLogItem `json:"sent_log,omitempty"`
	UpdatedAt       time.Time               `json:"updated_at"`
}

type AgentTopicProfile struct {
	ID        string             `json:"id"`
	Name      string             `json:"name"`
	Sources   []AgentTopicSource `json:"sources"`
	UpdatedAt time.Time          `json:"updated_at"`
}

type AgentTopicSource struct {
	ID       string  `json:"id"`
	Name     string  `json:"name"`
	Category string  `json:"category"`
	FeedURL  string  `json:"feed_url"`
	Weight   float64 `json:"weight"`
	Enabled  bool    `json:"enabled"`
}

type AgentTopicRun struct {
	ID          string           `json:"id"`
	ProfileID   string           `json:"profile_id"`
	CreatedAt   time.Time        `json:"created_at"`
	Summary     string           `json:"summary"`
	Items       []AgentTopicItem `json:"items"`
	FetchErrors []AgentRunError  `json:"fetch_errors,omitempty"`
}

type AgentTopicItem struct {
	Title       string `json:"title"`
	URL         string `json:"url"`
	SourceID    string `json:"source_id"`
	SourceName  string `json:"source_name"`
	PublishedAt string `json:"published_at,omitempty"`
	Summary     string `json:"summary,omitempty"`
}

type AgentRunError struct {
	Target string `json:"target"`
	Error  string `json:"error"`
}

type AgentWritingSnapshot struct {
	Topics    []AgentWritingTopic `json:"topics"`
	UpdatedAt time.Time           `json:"updated_at"`
}

type AgentWritingTopic struct {
	ID               string                 `json:"id"`
	Title            string                 `json:"title"`
	Status           string                 `json:"status"`
	Materials        []AgentWritingMaterial `json:"materials"`
	State            AgentWritingState      `json:"state"`
	Backup           AgentWritingState      `json:"backup"`
	ArtifactRoot     string                 `json:"artifact_root,omitempty"`
	RawFiles         []AgentWritingRawFile  `json:"raw_files,omitempty"`
	Artifacts        AgentWritingArtifacts  `json:"artifacts,omitempty"`
	LastSummarizedAt *time.Time             `json:"last_summarized_at,omitempty"`
	CreatedAt        time.Time              `json:"created_at"`
	UpdatedAt        time.Time              `json:"updated_at"`
}

type AgentWritingMaterial struct {
	ID           string         `json:"id"`
	Title        string         `json:"title"`
	Content      string         `json:"content"`
	Type         string         `json:"type,omitempty"`
	Source       string         `json:"source,omitempty"`
	InputMode    string         `json:"input_mode,omitempty"`
	URLs         []string       `json:"urls,omitempty"`
	RawFile      string         `json:"raw_file,omitempty"`
	ArtifactPath string         `json:"artifact_path,omitempty"`
	Metadata     map[string]any `json:"metadata,omitempty"`
	CreatedAt    time.Time      `json:"created_at"`
}

type AgentWritingState struct {
	Summary string `json:"summary"`
	Outline string `json:"outline"`
	Draft   string `json:"draft"`
}

type AgentWritingRawFile struct {
	Name      string `json:"name"`
	Path      string `json:"path"`
	LineCount int    `json:"line_count"`
}

type AgentWritingArtifacts struct {
	MaterialCount  int                   `json:"material_count"`
	InsightCount   int                   `json:"insight_count"`
	DocumentCount  int                   `json:"document_count"`
	LatestInsight  *AgentWritingInsight  `json:"latest_insight,omitempty"`
	LatestDocument *AgentWritingDocument `json:"latest_document,omitempty"`
}

type AgentWritingInsight struct {
	ID           string    `json:"id"`
	TopicID      string    `json:"topic_id"`
	MaterialIDs  []string  `json:"material_ids"`
	Summary      string    `json:"summary"`
	KeyPoints    []string  `json:"key_points"`
	Tags         []string  `json:"tags"`
	Entities     []string  `json:"entities"`
	QualityScore float64   `json:"quality_score"`
	Path         string    `json:"path,omitempty"`
	CreatedAt    time.Time `json:"created_at"`
}

type AgentWritingDocument struct {
	ID          string    `json:"id"`
	TopicID     string    `json:"topic_id"`
	MaterialIDs []string  `json:"material_ids"`
	InsightID   string    `json:"insight_id"`
	Mode        string    `json:"mode"`
	Title       string    `json:"title"`
	Path        string    `json:"path"`
	Version     int       `json:"version"`
	CreatedAt   time.Time `json:"created_at"`
}

type AgentMarketSnapshot struct {
	Portfolio AgentMarketPortfolio `json:"portfolio"`
	Config    AgentMarketConfig    `json:"config"`
	Runs      []AgentMarketRun     `json:"runs"`
	UpdatedAt time.Time            `json:"updated_at"`
}

type AgentMarketPortfolio struct {
	Funds []AgentMarketHolding `json:"funds"`
	Cash  float64              `json:"cash"`
}

type AgentMarketHolding struct {
	Code     string  `json:"code"`
	Name     string  `json:"name"`
	Quantity float64 `json:"quantity,omitempty"`
	AvgCost  float64 `json:"avg_cost,omitempty"`
}

type AgentMarketConfig struct {
	AnalysisEngine string `json:"analysis_engine"`
	SearchEngine   string `json:"search_engine"`
}

type AgentMarketRun struct {
	ID          string                    `json:"id"`
	CreatedAt   time.Time                 `json:"created_at"`
	Phase       string                    `json:"phase"`
	MarketState string                    `json:"market_state"`
	Summary     string                    `json:"summary"`
	Images      []AgentMarkdownImage      `json:"images,omitempty"`
	Assets      []AgentMarketAssetContext `json:"assets,omitempty"`
	SourceChain []string                  `json:"source_chain,omitempty"`
	Errors      []AgentRunError           `json:"errors,omitempty"`
}

type AgentEvolutionSnapshot struct {
	Goals     []AgentEvolutionGoal `json:"goals"`
	UpdatedAt time.Time            `json:"updated_at"`
}

type AgentEvolutionGoal struct {
	ID              string                     `json:"id"`
	Goal            string                     `json:"goal"`
	CommitMessage   string                     `json:"commit_message,omitempty"`
	Status          string                     `json:"status"`
	Stage           string                     `json:"stage"`
	Plan            AgentEvolutionPlan         `json:"plan"`
	FixAttempts     int                        `json:"fix_attempts"`
	StartedFromRef  string                     `json:"started_from_ref,omitempty"`
	CompletedCommit string                     `json:"completed_commit,omitempty"`
	TestResults     []AgentEvolutionTestResult `json:"test_results,omitempty"`
	RawTail         []AgentEvolutionRawLine    `json:"raw_tail,omitempty"`
	LastCodexOutput string                     `json:"last_codex_output,omitempty"`
	Events          []AgentEvolutionEvent      `json:"events"`
	CreatedAt       time.Time                  `json:"created_at"`
	UpdatedAt       time.Time                  `json:"updated_at"`
	StartedAt       *time.Time                 `json:"started_at,omitempty"`
	CompletedAt     *time.Time                 `json:"completed_at,omitempty"`
	LastError       string                     `json:"last_error,omitempty"`
}

type AgentEvolutionPlan struct {
	Steps       []string `json:"steps"`
	CurrentStep int      `json:"current_step"`
}

type AgentEvolutionTestResult struct {
	Name       string    `json:"name"`
	Command    string    `json:"command"`
	OK         bool      `json:"ok"`
	ExitCode   int       `json:"exit_code"`
	Output     string    `json:"output"`
	StartedAt  time.Time `json:"started_at"`
	FinishedAt time.Time `json:"finished_at"`
}

type AgentEvolutionRawLine struct {
	At   time.Time `json:"at"`
	Line string    `json:"line"`
}

type AgentEvolutionEvent struct {
	At      time.Time `json:"at"`
	Stage   string    `json:"stage"`
	Message string    `json:"message"`
}

type AgentTerminalRequest struct {
	Command string `json:"command"`
	CWD     string `json:"cwd,omitempty"`
}

type AgentTerminalResult struct {
	Command    string    `json:"command"`
	CWD        string    `json:"cwd,omitempty"`
	ExitCode   int       `json:"exit_code"`
	Output     string    `json:"output"`
	StartedAt  time.Time `json:"started_at"`
	FinishedAt time.Time `json:"finished_at"`
}

File: web/admin/src/lib/agent.ts
import { request } from './api';

export type AgentLLMProvider = {
  id: string;
  name: string;
  type: string;
  base_url?: string;
  api_key?: string;
  model?: string;
  chat_path?: string;
  timeout_ms?: number;
};

export type AgentSettings = {
  runtime_mode: string;
  default_llm_provider_id: string;
  llm_providers: AgentLLMProvider[];
  wecom: {
    corp_id?: string;
    corp_secret?: string;
    agent_id?: string;
    base_url?: string;
    bridge_url?: string;
    bridge_token?: string;
    bridge_stream_enabled?: boolean;
    audio_dir?: string;
    text_max_bytes?: number;
    enabled: boolean;
  };
  terminal: {
    enabled: boolean;
    cwd?: string;
    timeout_ms?: number;
  };
  evolution: {
    command?: string;
    cwd?: string;
    timeout_ms?: number;
    codex_model?: string;
    codex_reasoning?: string;
    max_fix_attempts?: number;
    test_commands?: Array<{ name: string; command: string; timeout_ms?: number }>;
    auto_commit?: boolean;
    auto_push?: boolean;
    push_remote?: string;
    push_branch?: string;
    structure_review?: boolean;
  };
  stt?: Record<string, unknown>;
  search_engines?: Array<Record<string, unknown>>;
  memory?: Record<string, unknown>;
  md2img?: Record<string, unknown>;
  updated_at: string;
};

export type AgentDirectInputRule = {
  id: string;
  name: string;
  pattern: string;
  target_text: string;
  match_mode: 'exact' | 'fuzzy';
  enabled: boolean;
};

export type AgentDirectInputConfig = {
  version: number;
  rules: AgentDirectInputRule[];
  updated_at: string;
};

export type AgentWeComButton = {
  id: string;
  name: string;
  key: string;
  enabled: boolean;
  dispatch_text: string;
  sub_buttons?: AgentWeComButton[];
};

export type AgentWeComMenuConfig = {
  version: number;
  buttons: AgentWeComButton[];
  updated_at: string;
  last_published_at?: string | null;
};

export type AgentWeComMenuSnapshot = {
  config: AgentWeComMenuConfig;
  recent_events: Array<Record<string, unknown>>;
  publish_payload?: Record<string, unknown> | null;
  validation_errors?: string[];
};

export type AgentWeComUser = {
  id: string;
  name: string;
  wecom_user: string;
  enabled: boolean;
  updated_at?: string;
};

export type AgentPushSnapshot = {
  users: AgentWeComUser[];
  updated_at: string;
};

export type AgentConversation = {
  id: string;
  session_id: string;
  input: string;
  resolved?: string;
  response: string;
  status: string;
  metadata?: Record<string, unknown>;
  created_at: string;
};

export type AgentCapabilityInfo = {
  name: string;
  description: string;
  terminal?: boolean;
  command?: string;
  install?: string;
  keywords?: string[];
  direct_commands?: string[];
  tool?: string;
  action?: string;
  params?: string[];
  prefer_tool_result?: boolean;
  detail?: string;
};

export type AgentCapabilityRunResult = {
  capability: string;
  tool?: string;
  action?: string;
  input?: string;
  used_command?: string;
  output?: string;
  terminal?: TerminalResult;
  metadata?: Record<string, unknown>;
};

export type AgentMarkdownRenderResult = {
  mode: string;
  images: Array<{ path: string; content_type: string; size_bytes: number; width?: number; height?: number }>;
  output_dir: string;
  source_chars: number;
  rendered_at: string;
};

export type AgentTopicSnapshot = {
  active_profile_id: string;
  profiles: Array<Record<string, unknown>>;
  runs: Array<Record<string, unknown>>;
  updated_at: string;
};

export type AgentWritingTopic = {
  id: string;
  title: string;
  status: string;
  materials: Array<Record<string, unknown>>;
  state: Record<string, string>;
  backup: Record<string, string>;
  artifact_root?: string;
  raw_files?: Array<Record<string, unknown>>;
  artifacts?: Record<string, unknown>;
  last_summarized_at?: string | null;
  created_at: string;
  updated_at: string;
};

export type AgentMarketPortfolio = {
  funds: Array<{
    code: string;
    name: string;
    quantity?: number;
    avg_cost?: number;
  }>;
  cash: number;
};

export type AgentMarketSnapshot = {
  portfolio: AgentMarketPortfolio;
  config: Record<string, unknown>;
  runs: Array<Record<string, unknown>>;
  updated_at: string;
};

export type AgentEvolutionGoal = {
  id: string;
  goal: string;
  commit_message?: string;
  status: string;
  stage: string;
  plan?: { steps: string[]; current_step: number };
  fix_attempts?: number;
  started_from_ref?: string;
  completed_commit?: string;
  test_results?: Array<Record<string, unknown>>;
  raw_tail?: Array<Record<string, unknown>>;
  last_codex_output?: string;
  events: Array<Record<string, unknown>>;
  created_at: string;
  updated_at: string;
  last_error?: string;
};

export type AgentSnapshot = {
  settings: AgentSettings;
  capabilities: AgentCapabilityInfo[];
  direct_input: AgentDirectInputConfig;
  wecom_menu: AgentWeComMenuSnapshot;
  push: AgentPushSnapshot;
  conversations: AgentConversation[];
  topic_summary: AgentTopicSnapshot;
  writing: {
    topics: AgentWritingTopic[];
    updated_at: string;
  };
  market: AgentMarketSnapshot;
  evolution: {
    goals: AgentEvolutionGoal[];
    updated_at: string;
  };
  updated_at: string;
};

export type TerminalResult = {
  command: string;
  cwd?: string;
  exit_code: number;
  output: string;
  started_at: string;
  finished_at: string;
};

export type AgentSearchResult = {
  items: Array<{
    title: string;
    source: string;
    link?: string;
    published_at?: string;
    snippet?: string;
  }>;
  source_chain: string[];
  errors: string[];
};

export function fetchAgentSnapshot() {
  return request<AgentSnapshot>('/agent').then(normalizeAgentSnapshot);
}

export function saveAgentSettings(payload: AgentSettings) {
  return request<AgentSnapshot>('/agent/settings', { method: 'PUT', body: JSON.stringify(payload) }).then(normalizeAgentSnapshot);
}

export function saveAgentDirectInput(payload: AgentDirectInputConfig) {
  return request<AgentSnapshot>('/agent/direct-input', { method: 'PUT', body: JSON.stringify(payload) }).then(normalizeAgentSnapshot);
}

export function saveAgentPush(payload: AgentPushSnapshot) {
  return request<AgentSnapshot>('/agent/push', { method: 'PUT', body: JSON.stringify(payload) }).then(normalizeAgentSnapshot);
}

export function saveAgentWeComMenu(payload: AgentWeComMenuConfig) {
  return request<AgentSnapshot>('/agent/wecom/menu', { method: 'PUT', body: JSON.stringify(payload) }).then(normalizeAgentSnapshot);
}

export function publishAgentWeComMenu() {
  return request<AgentWeComMenuSnapshot>('/agent/wecom/menu/publish', { method: 'POST' });
}

export function sendAgentWeComMessage(payload: { to_user: string; text: string }) {
  return request<{ ok: boolean }>('/agent/wecom/send', { method: 'POST', body: JSON.stringify(payload) });
}

export function sendAgentWeComImage(payload: { to_user: string; base64: string; filename?: string; content_type?: string }) {
  return request<{ ok: boolean }>('/agent/wecom/image', { method: 'POST', body: JSON.stringify(payload) });
}

export function runAgentConversation(payload: { input: string; session_id?: string }) {
  return request<AgentConversation>('/agent/conversation', { method: 'POST', body: JSON.stringify(payload) });
}

export function fetchAgentCapabilities() {
  return request<AgentCapabilityInfo[]>('/agent/capabilities').then((items) => (Array.isArray(items) ? items : []));
}

export function fetchAgentCapability(name: string) {
  return request<AgentCapabilityInfo>(`/agent/capabilities/${encodeURIComponent(name)}`);
}

export function runAgentCapability(name: string, payload: { input?: string; command?: string; args?: string[] }) {
  return request<AgentCapabilityRunResult>(`/agent/capabilities/${encodeURIComponent(name)}/run`, {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export function saveAgentTopic(payload: AgentTopicSnapshot) {
  return request<AgentSnapshot>('/agent/topic', { method: 'PUT', body: JSON.stringify(payload) }).then(normalizeAgentSnapshot);
}

export function runAgentTopic(profileId?: string) {
  return request<Record<string, unknown>>('/agent/topic/run', {
    method: 'POST',
    body: JSON.stringify({ profile_id: profileId ?? '' }),
  });
}

export function createWritingTopic(payload: { id?: string; title: string }) {
  return request<AgentWritingTopic>('/agent/writing/topics', { method: 'POST', body: JSON.stringify(payload) });
}

export function addWritingMaterial(topicId: string, payload: { title?: string; content: string }) {
  return request<AgentWritingTopic>(`/agent/writing/topics/${encodeURIComponent(topicId)}/materials`, {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export function summarizeWritingTopic(topicId: string) {
  return request<AgentWritingTopic>(`/agent/writing/topics/${encodeURIComponent(topicId)}/summarize`, { method: 'POST' });
}

export function saveMarketPortfolio(payload: AgentMarketPortfolio) {
  return request<AgentSnapshot>('/agent/market/portfolio', { method: 'PUT', body: JSON.stringify(payload) }).then(normalizeAgentSnapshot);
}

export function importMarketPortfolioCodes(payload: { codes: string }) {
  return request<Record<string, unknown>>('/agent/market/portfolio/import-codes', { method: 'POST', body: JSON.stringify(payload) });
}

export function runMarketAnalysis(payload: { phase: string; notes?: string }) {
  return request<Record<string, unknown>>('/agent/market/run', { method: 'POST', body: JSON.stringify(payload) });
}

export function createEvolutionGoal(payload: { goal: string; commit_message?: string }) {
  return request<AgentEvolutionGoal>('/agent/evolution/goals', { method: 'POST', body: JSON.stringify(payload) });
}

export function runEvolutionGoal(goalId: string) {
  return request<AgentEvolutionGoal>(`/agent/evolution/goals/${encodeURIComponent(goalId)}/run`, { method: 'POST' });
}

export function runTerminalCommand(payload: { command: string; cwd?: string }) {
  return request<TerminalResult>('/agent/terminal', { method: 'POST', body: JSON.stringify(payload) });
}

export function runAgentSearch(payload: {
  engine_selector?: string;
  timeout_ms?: number;
  max_items?: number;
  plans: Array<{ label: string; query: string; sites?: string[]; recency?: string }>;
}) {
  return request<AgentSearchResult>('/agent/search/run', { method: 'POST', body: JSON.stringify(payload) });
}

export function transcribeAgentSpeech(payload: { audio_path: string }) {
  return request<Record<string, unknown>>('/agent/stt/transcribe', { method: 'POST', body: JSON.stringify(payload) });
}

export function runAgentCodex(payload: {
  task_id?: string;
  prompt: string;
  model?: string;
  reasoning_effort?: string;
  timeout_ms?: number;
  cwd?: string;
}) {
  return request<Record<string, unknown>>('/agent/codex/run', { method: 'POST', body: JSON.stringify(payload) });
}

export function renderAgentMarkdown(payload: { markdown: string; mode?: 'long-image' | 'multi-page'; output_dir?: string }) {
  return request<AgentMarkdownRenderResult>('/agent/md2img/render', { method: 'POST', body: JSON.stringify(payload) });
}

export function stableJSON(value: unknown) {
  return JSON.stringify(value, null, 2);
}

export function normalizeAgentSnapshot(input: AgentSnapshot): AgentSnapshot {
  const snapshot = input ?? ({} as AgentSnapshot);
  const settings = snapshot.settings ?? ({} as AgentSettings);
  const directInput = snapshot.direct_input ?? ({} as AgentDirectInputConfig);
  const wecomMenu = snapshot.wecom_menu ?? ({} as AgentWeComMenuSnapshot);
  const wecomConfig = wecomMenu.config ?? ({} as AgentWeComMenuConfig);
  const push = snapshot.push ?? ({} as AgentPushSnapshot);
  const topic = snapshot.topic_summary ?? ({} as AgentTopicSnapshot);
  const market = snapshot.market ?? ({} as AgentMarketSnapshot);
  const portfolio = market.portfolio ?? ({ cash: 0 } as AgentMarketPortfolio);
  const writing = snapshot.writing ?? { topics: [], updated_at: '' };
  const evolution = snapshot.evolution ?? { goals: [], updated_at: '' };

  return {
    ...snapshot,
    settings: {
      ...settings,
      llm_providers: arrayOrEmpty(settings.llm_providers),
      search_engines: arrayOrEmpty(settings.search_engines),
      wecom: settings.wecom ?? { enabled: false },
      terminal: settings.terminal ?? { enabled: false },
      evolution: {
        ...(settings.evolution ?? {}),
        test_commands: arrayOrEmpty(settings.evolution?.test_commands),
      },
    },
    capabilities: arrayOrEmpty(snapshot.capabilities),
    direct_input: {
      ...directInput,
      rules: arrayOrEmpty(directInput.rules),
    },
    wecom_menu: {
      ...wecomMenu,
      config: {
        ...wecomConfig,
        buttons: arrayOrEmpty(wecomConfig.buttons),
      },
      recent_events: arrayOrEmpty(wecomMenu.recent_events),
      validation_errors: arrayOrEmpty(wecomMenu.validation_errors),
    },
    push: {
      ...push,
      users: arrayOrEmpty(push.users),
    },
    conversations: arrayOrEmpty(snapshot.conversations),
    topic_summary: {
      ...topic,
      profiles: arrayOrEmpty(topic.profiles),
      runs: arrayOrEmpty(topic.runs),
    },
    writing: {
      ...writing,
      topics: arrayOrEmpty(writing.topics),
    },
    market: {
      ...market,
      portfolio: {
        ...portfolio,
        funds: arrayOrEmpty(portfolio.funds),
        cash: typeof portfolio.cash === 'number' ? portfolio.cash : 0,
      },
      runs: arrayOrEmpty(market.runs),
    },
    evolution: {
      ...evolution,
      goals: arrayOrEmpty(evolution.goals),
    },
  };
}

function arrayOrEmpty<T>(value: T[] | null | undefined): T[] {
  return Array.isArray(value) ? value : [];
}
